File: Problems/0112_Path_Sum/Path_Sum.cs
using System;

// Definition for a binary tree node.
public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) {
        val = x;
    }
}

public class Solution {
    public bool HasPathSum(TreeNode root, int sum)
    {
        if (root == null) {
            Console.WriteLine("root.val = null, " + "sum =  " + sum.ToString());
        }
        else {
            Console.WriteLine("root.val = " + root.val.ToString() + ", sum = " + sum.ToString());
        }

        if (root == null) {
            return false;
        }

        if ((root.left == null) && (root.right == null)) {
            if (root.val == sum) {
                Console.WriteLine("root.val = " + root.val.ToString() + ", root.left = null, root.right = null, sum= " + sum.ToString() + ", return(true)");
                return true;
            }
            else {
                Console.WriteLine("root.val = " + root.val.ToString() + ", root.left = null, root.right = null, sum= " + sum.ToString() + ", return(false)");
                return false;
            }
        }
        if (root.left != null) {
            if (HasPathSum(root.left, sum - root.val)) {
                return true;
            }
        }
        if (root.right != null) {
            if (HasPathSum(root.right, sum - root.val)) {
                return true;
            }
        }

        return false;
    }

    private void load_treeData(TreeNode root, string data)
    {
        if (data.Length == 0)
            Console.WriteLine("load_treeData() ... arg data Error");
        
        string[] args = data.Split(',');
        if (args[0] != "null") {

        }

        root.val = data[0];
    }

    private TreeNode load_sample0_treeData()
    {
        TreeNode root = new TreeNode(5);
        root.left = new TreeNode(4);
        root.right = new TreeNode(8);
        root.left.left = new TreeNode(11);
        root.left.left.left = new TreeNode(7);
        root.left.left.right = new TreeNode(2);
        root.right.left = new TreeNode(13);
        root.right.right = new TreeNode(4);
        root.right.right.right = new TreeNode(1);

        return root;
    }

    private TreeNode load_sample2_treeData()
    {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);

        return root;
    }

    public void Main(string args)
    {
    //    Console.WriteLine("args = " + args);    
    //    string s = args.Replace("\"", "");

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

    //    TreeNode root = load_sample0_treeData();
    //    Console.WriteLine("Result = " + HasPathSum(root, 22).ToString());

        TreeNode root = load_sample2_treeData();
        Console.WriteLine("Result = " + HasPathSum(root, 1).ToString());
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms");
    }
}

File: Problems/0112_Path_Sum/Path_Sum.py
# coding: utf-8

import sys
import time

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        if root == None:
            return False

        if root.left == None and root.right == None:
            if root.val == sum:
                return True
            else:
                return False

        if root.left != None:
            if self.hasPathSum(root.left, sum - root.val):
                return True

        if root.right != None:
            if self.hasPathSum(root.right, sum - root.val):
                return True

        return False


def load_sample0_treeData():
    root = TreeNode(5)
    root.left = TreeNode(4)
    root.right = TreeNode(8)
    root.left.left = TreeNode(11)
    root.left.left.left = TreeNode(7)
    root.left.left.right = TreeNode(2)
    root.right.left = TreeNode(13)
    root.right.right = TreeNode(4)
    root.right.right.right = TreeNode(1)
    return root

def load_sample2_treeData():
    root = TreeNode(1)
    root.left = TreeNode(2)
    return root


def main():
    time0 = time.time()

    sl = Solution()
    '''
    root = load_sample0_treeData()
    print(sl.hasPathSum(root, 22))
    '''
    root = load_sample2_treeData()
    print(sl.hasPathSum(root, 1))

    time1 = time.time()
    print("Execute time ... : %f[s]" %(time1 - time0))
    print()


if __name__ == "__main__":
    main()
